sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriter.java

```diff
@@ -56,15 +55,18 @@ public void processElement(ProcessContext ctx) throws Exception {
     checkForFailures();
 
     ProducerRecord<K, V> record = ctx.element();
-    KV<K, V> kv = KV.of(record.key(), record.value());
-
     Long timestampMillis =
-        spec.getPublishTimestampFunction() != null
-            ? spec.getPublishTimestampFunction().getTimestamp(record, ctx.timestamp()).getMillis()
-            : null;
+        record.timestamp() != null
+            ? record.timestamp()
+            : (spec.getPublishTimestampFunction() != null
+                ? spec.getPublishTimestampFunction()
+                    .getTimestamp(record, ctx.timestamp())
+                    .getMillis()
+                : System.currentTimeMillis());
```

Comment:
No need to use currentTimeMillis(). We should leave it 'null' such that Kafka client decide.

Reply:
Agree, fixed